client/my-sites/stats/stats-date-picker/index.jsx

```diff
@@ -56,16 +56,43 @@ class StatsDatePicker extends Component {
 		}
 	}
 
-	dateForCustomRange() {
-		const { query, moment, translate } = this.props;
+	dateForCustomRange( startDate, endDate ) {
+		const { moment, translate } = this.props;
+
+		const localizedStartDate = moment( startDate );
+		const localizedEndDate = moment( endDate );
+
+		// If it's the same day, show single date.
+		if ( localizedStartDate.isSame( localizedEndDate, 'day' ) ) {
+			return localizedStartDate.format( 'LL' );
+		}
+
+		// If it's a full month.
+		if (
+			localizedStartDate.isSame( localizedStartDate.clone().startOf( 'month' ), 'day' ) &&
+			localizedEndDate.isSame( localizedEndDate.clone().endOf( 'month' ), 'day' ) &&
+			localizedStartDate.isSame( localizedEndDate, 'month' )
+		) {
+			return localizedStartDate.format( 'MMMM YYYY' );
+		}
+
+		// If it's a full year.
+		if (
+			localizedStartDate.isSame( localizedStartDate.clone().startOf( 'year' ), 'day' ) &&
+			localizedEndDate.isSame( localizedEndDate.clone().endOf( 'year' ), 'day' ) &&
+			localizedStartDate.isSame( localizedEndDate, 'year' )
+		) {
+			return localizedStartDate.format( 'YYYY' );
+		}
 
-		const localizedStartDate = moment( query.start_date );
-		const localizedEndDate = moment( query.date );
+		// Default to date range
+		const firstFormatString =
+			localizedStartDate.year() === localizedEndDate.year() ? 'MMM D' : 'll';
 
 		return translate( '%(startDate)s ~ %(endDate)s', {
 			context: 'Date range for which stats are being displayed',
 			args: {
-				startDate: localizedStartDate.format( 'll' ),
+				startDate: localizedStartDate.format( firstFormatString ),
 				endDate: localizedEndDate.format( 'll' ),
```

Comment:
We might want to add the year after string 'MMM D', otherwise it looks broken for locales that use year to start the date:

![image](https://github.com/user-attachments/assets/0832274c-e573-4334-8954-f551fb3fe852)

Reply:
https://github.com/Automattic/wp-calypso/pull/96835/commits/22a4d93c99f5f53ae0e9f4b24d377feca92dc6f8